Implement cross-correlation redshift/rv-fiding (#544)
astropy/specutils: docs/analysis.rst

   >>> tm_result = template_comparison.template_match(observed_spectrum, spectral_template) # doctest:+FLOAT_CMP

> Cross-correlation

Reviewer: ```suggestion
Template Cross-correlation
```

-----------------

The cross-correlation function between an observed spectrum and a template spectrum that both share a common spectral

> axis, can be calculated by function `~template_correlate` in the `specutils.analysis.correlation` module.

Reviewer: ```suggestion
axis can be calculated with the function `~template_correlate` in the `specutils.analysis.correlation` module.
```

An example of how to get the cross correlation follows. Note that the observed spectrum must have a rest wavelength

> value set in.

Reviewer: ```suggestion
value set.
```

> .. code-block:: python

Reviewer: Probably a question for @eteq or @camipacifici: would be want to demonstrate loading in a simple ASCII file with e.g. wavelength/flux values to use as the template? (That is, for someone perusing the docs, would that be more of a helpful example?)

Reviewer 2: Hmm, I think the more useful one is something that actually uses *real* templates that are loaded as `Spectrum1D` objects.  That is, a fits file with an actual spectrum.

Reviewer: E.g. take a galaxy template from synphot and convert it to ascii and load in into a Spectrum1D object to use as the template.

Reviewer 3: There are some templates in the Synphot library (Bruzual-Charlot galaxy templates for example) https://pysynphot.readthedocs.io/en/latest/ $PYSYN_CDBS/grid/bc95/. I'd suggest bc95_g_10E9.fits. I don't think there is a URL available for this single file, but @pllim might know.

Author: In developing a test case which uses a real template spectrum (wave and flux from bc95_g_10E9.fits converted to ASCII format), I realized that we need also an observed spectrum to correlate against this template. Any idea on how to get one of those without resorting to too many manipulations? The goal is to have this in the docs, so I believe it has to be straightforward to get the observed spectrum, as it is to get the template.

Reviewer 3: Maybe grab a template that is close to it in age (bc95_g_50E8.fits) and redshift it and add a little Gaussian noise?

Reviewer 2: @ibusko - For an observed spectrum, I think the easiest thing is to use the SDSS spectrum that's already an example from the specutils docs - see https://specutils.readthedocs.io/en/stable/ (or the `index.rst` here) for the exact URL. That way we don't have to keep track of multiple examples.

Reviewer 2: Although @camipacifici pointed out to me out-of-band that she has some actual observed spectra and matched templates for them that is part of the notebook she's working on. Maybe we should use those as the template/observed combo?

Unfortunately I'm not sure it's going to be easy to access these publicly until she gets a chance to put them in a reasonable data location. So it might be best to follow the above just to get this PR in, but with a plan to move to @camipacifici's data once we can get it (and use it also in the chi2-template-fitting section?

Reviewer 4: The data is served by RedCat, independent of Synphotses code base. Pick your poison at https://ssb.stsci.edu/cdbs/grid/bc95/templates/

Author: I put in place a unit test function in a branch in my fork for people to review and continue the discussion. The goal is to have something that can be put in the docs as an example on how to use the correlation function with real data. I found that it is perhaps trickier and lengthier than it should be, since it involves other aspects of the process besides the correlation itself. The code is in function ` test_correlation_real_data` in  [ibusko:redshift-measurement-sandbox](https://github.com/ibusko/specutils/blob/redshift-measurement-sandbox/specutils/tests/test_correlation.py).

Temporarily I had both the observation and template spectra files added to the `tests` directory, because my dev environment  (in PyCharm) can't run the test (it skips it) if I use the remote file loading code. Another point is that remote loading works for the observed SDSS spectrum, but not for the synphot template. This file has to be manually changed into ECSV format to be usable. These are topics to be solved later on. For now, we have to decide how to translate the code in the test function into a doc-usable code snippet.

Reviewer 4: FWIW, I implemented API for `synphot` to read from and write to `Spectrum1D`. Maybe it'll make things easier for you. See spacetelescope/synphot_refactor#243

   >>> from specutils.analysis import correlation
   >>> from specutils.spectra.spectrum1d import Spectrum1D
   >>> ospec = Spectrum1D(spectral_axis=..., flux=..., ..., \
                          velocity_convention='optical', rest_value=...)
   >>> tspec = Spectrum1D(spectral_axis=..., flux=..., ...)
   >>> corr, lag = correlation.template_correlate(ospec, tspec)

The lag values are reported in km/s units. The correlation values are computed after the template spectrum is

> normalized such as to have the same total flux as the observed spectrum.

Reviewer: ```suggestion
normalized in order to have the same total flux as the observed spectrum.
```
